### PR TITLE
Modified work with parameters and config files

### PR DESCRIPTION
**Pull request by Dimentium**
Hi,
I'm not software developer, and absolutely new to Obj-C.
But tried to do my best :)

### CODE REVIEW

File: AutoRaise.mm [hunk 1]
```diff
@@ -475,48 +475,140 @@ CGEventRef eventTapHandler(CGEventTapProxy proxy, CGEventType type, CGEventRef e
...
+        #define defaultWarpY @"warpY"
+        #define defaultScale @"scale"
+        
+        // initial values for standardDefaults
+        NSDictionary* standardDefaults = @{defaultDelay:@-1,
+                                           defaultWarpX:@-1.0,
+                                           defaultWarpY:@-1.0,
+                                           defaultScale:@2.0 };
+        
+        // sync standardDefaults with command-line arguments
```
Comment (sbmpost) [inline AutoRaise.mm:494]: This seems like a better way to set the defaults when compared to what I am doing (reading the fields first and then provide default values for them if we got nil/null objects). There are actually 4 sources of information to consider:

1 command line parameters
2 original file structure
3 new file structure
4 defaults

If we could apply this sync strategy in a smart way so that we respect the priority of 1 to 4, then the code will probably be cleaner indeed.

File: AutoRaise.mm [hunk 2]
```diff
@@ -475,48 +475,140 @@ CGEventRef eventTapHandler(CGEventTapProxy proxy, CGEventType type, CGEventRef e
...
+
+        // if there are any arguments, we would use only them
+        if (argc > 1) {
+            // but if -delay argument not valid, we would exit
+            if (not (delayCount > 0)) {
+                fprintf(stderr, "No valid arguments\nRun without any arguments to use default settings");
+                return 0;
             }
+            // check is there valid warps
+            if (warpX >= 0 and warpX <= 1 and warpY >= 0 and warpY <= 1) { warpMouse = true; }
```
Comment (sbmpost) [inline AutoRaise.mm:511]: So you actually chose to implement validation checks. That is cool, but also makes the code quite a bit more complex. So I myself chose to not do any validation checks at all (and instead print out clearly to the command line what settings are being applied).

File: AutoRaise.mm [hunk 3]
```diff
@@ -475,48 +475,140 @@ CGEventRef eventTapHandler(CGEventTapProxy proxy, CGEventType type, CGEventRef e
...
-                NSString * line = [[NSString alloc] initWithData:
-                    [warpFile readDataOfLength:11] encoding:
-                    NSUTF8StringEncoding];
-                NSArray * components = [line componentsSeparatedByString: @" "];
-                if (components.count >= 1) { warpX = [[components objectAtIndex:0] floatValue]; }
-                if (components.count >= 2) { warpY = [[components objectAtIndex:1] floatValue]; }
-                if (components.count >= 3) { cursorScale = [[components objectAtIndex:2] floatValue]; }
-                [warpFile closeFile];
+            
+            NSString * oldStyleDelayFile   = [NSString stringWithFormat: @"%@/AutoRaise.delay",          home];
```
Comment (sbmpost) [inline AutoRaise.mm:516]: ah, there we go, the requested feature :-)

File: AutoRaise.mm [hunk 4]
```diff
@@ -475,48 +475,140 @@ CGEventRef eventTapHandler(CGEventTapProxy proxy, CGEventType type, CGEventRef e
...
+            NSString * oldStyleWarpFile    = [NSString stringWithFormat: @"%@/AutoRaise.warp",           home];
+            NSString * newStyleConfigFile1 = [NSString stringWithFormat: @"%@/.AutoRaise",               home];
+            NSString * newStyleConfigFile2 = [NSString stringWithFormat: @"%@/.config/AutoRaise/config", home];
+            
+            #define isFileExist(string) [[NSFileManager defaultManager] fileExistsAtPath:string]
```
Comment (sbmpost) [inline AutoRaise.mm:521]: You can also turn this into a method. #define also works, but is more intended for defining constants. Not sure if objective-c supports the __inline modifier on methods.

File: AutoRaise.mm [hunk 5]
```diff
@@ -475,48 +475,140 @@ CGEventRef eventTapHandler(CGEventTapProxy proxy, CGEventType type, CGEventRef e
...
+            
+            // old style config files
```
Comment (sbmpost) [inline AutoRaise.mm:523]: I would move the old and new style code into their own methods, just to keep things more understandable.

File: AutoRaise.mm [hunk 6]
```diff
@@ -475,48 +475,140 @@ CGEventRef eventTapHandler(CGEventTapProxy proxy, CGEventType type, CGEventRef e
...
+                    NSArray  * newStyleConfigLines = [newStyleConfig componentsSeparatedByString:@"\n"];
+
+                    NSString * newStyleConfigLine;
+                    NSArray  * temporaryConfigDictionary = @[@"delay", @"warpx", @"warpy", @"scale"];
+                    int newStyleConfigDelay = -1;
+                    float newStyleConfigWarpX = -1.0;
+                    float newStyleConfigWarpY = -1.0;
+                    float newStyleConfigScale = -1.0;
+
+                    for ( int i = 0; i < [newStyleConfigLines count]; i++) {
```
Comment (sbmpost) [inline AutoRaise.mm:572]: ah, reading the new format here, line by line. probably worth of its own method as well ;-)

**Review by sbmpost — COMMENTED** [review]
@Dimentium I made some comments :-)

File: README.md [hunk 7]
```diff
@@ -28,6 +28,14 @@ file. In order to pass the parameters from the example above by means of these c
     echo 1 > ~/AutoRaise.delay
     echo "0.5 0.1 2.5" > ~/AutoRaise.warp
 
+Using of **~/.AutoRaise** or **~/.config/AutoRaise/config** also possible:
```
Comment (sbmpost) [inline README.md:31]: Good that you pay intention to this detail :-)
Might want to rephrase it to something like:

"Update (2021-04-16): In addition to the configuration files mentioned above, AutoRaise now supports hidden configuration files in these locations: **~/.AutoRaise** or **~/.config/AutoRaise/config**. The format of these files is different as well:

File: AutoRaise.mm (hunk 1 above)
Comment (Dimentium) [reply AutoRaise.mm:494]: there are some issues with detecting nil objects in standardDefaults - its all have value = 0, instead nil  :/
so I set definitely wrong initial value (-1) and it is now clear - is there any arguments from command line was received.

Priority of arguments same as you wrote:
1) valid command line arguments
2) original file structure
3) new files structure
4) dafults 

I'm sorry for my bad code :(

File: README.md [hunk 8]
```diff
@@ -28,6 +28,14 @@ file. In order to pass the parameters from the example above by means of these c
...
+
+    #AutoRaise config file
+    delay 1
```
Comment (sbmpost) [inline README.md:34]: Usually key value pairs look like this:
delay=1
warpX=0.5
warpY=0.3
scale=1.8

File: AutoRaise.mm (hunk 1 above)
Comment (sbmpost) [reply AutoRaise.mm:494]: Code is not too bad. In fact I also struggled with checking for the "existence of a certain key in the standardDefaults". I think you'd first need to make it an NSDictionary (standardDefaults has a method for that), and then you will have more flexibility.

File: AutoRaise.mm (hunk 2 above)
Comment (Dimentium) [reply AutoRaise.mm:511]: it is not real validation.
this condition only for detecting warpMouse status.
I wanted do it  instead checking arguments quantity - otherwise false positive 'true' value possible sometimes.

File: AutoRaise.mm (hunk 4 above)
Comment (Dimentium) [reply AutoRaise.mm:521]: Ok. I need to learn something about 'methods'.

File: README.md (hunk 8 above)
Comment (Dimentium) [reply README.md:34]: Ok

File: README.md (hunk 7 above)
Comment (Dimentium) [reply README.md:31]: Ok

File: AutoRaise.mm (hunk 6 above)
Comment (Dimentium) [reply AutoRaise.mm:572]: Ok. I need to learn something about 'methods' first :-)

File: AutoRaise.mm (hunk 5 above)
Comment (Dimentium) [reply AutoRaise.mm:523]: Ok!

**Comment by Dimentium** [comment]
I totally rewrite my code with methods. 
But I'm not sure is it OK?

**Comment by sbmpost** [comment]
@Dimentium 
This already looks a lot better :)

I will be taking a more detailed look on Monday as the weekend is kind of reserved for the family ;-)

Nice to see how you are learning from this exercise so quickly :)

**Comment by Dimentium** [comment]
I think the same!
Have a nice weekend! :)

--- Dmitry Kuznetsov

File: AutoRaise.mm [hunk 9]
```diff
@@ -475,48 +599,30 @@ CGEventRef eventTapHandler(CGEventTapProxy proxy, CGEventType type, CGEventRef e
...
-                if (components.count >= 3) { cursorScale = [[components objectAtIndex:2] floatValue]; }
-                [warpFile closeFile];
+        printf("\nv%s by sbmpost(c) 2021, usage:\nAutoRaise -delay <1=%dms> [-warpX <0.5> -warpY <0.5> -scale <2.0>]", VERSION, POLLING_MS);
+        
+        ConfigClass * config = [[ConfigClass alloc]init];
+        
+        if (argc > 1) {                             // if there are any arguments, we would use only them
+            [config readCommandLineConfig];
+        } else {                                    // otherwise try to read
+            if (not [config readOriginalConfig]) {  // original config files
```
Comment (sbmpost) [inline AutoRaise.mm:612]: This logic probably disappears if you decide to merge NSDictionaries.

**Review by sbmpost — COMMENTED** [review]
Added some more detailed comments.

File: AutoRaise.mm [hunk 10]
```diff
@@ -334,6 +334,130 @@ - (void)onTick:(NSNumber *)timerInterval {
     oldPoint.x = oldPoint.y = 0;
 }
 
+//----------------------------------------------configuration-----------------------------------------------
+
+NSMutableDictionary *parameters = [[NSMutableDictionary alloc] init];
+#define _delay parameters[@"delay"]
```
[the remark this answers is not in the document]
Comment (Dimentium) [reply AutoRaise.mm:340]: Hm. I used them just as eye-candy thing.  
I understand that pre-processor replace them with right string.
I shouldn't do that?

File: AutoRaise.mm [hunk 11]
```diff
@@ -334,6 +334,130 @@ - (void)onTick:(NSNumber *)timerInterval {
     oldPoint.x = oldPoint.y = 0;
 }
 
+//----------------------------------------------configuration-----------------------------------------------
+
+NSMutableDictionary *parameters = [[NSMutableDictionary alloc] init];
```
[the remark this answers is not in the document]
Comment (Dimentium) [reply AutoRaise.mm:339]: I thought about it, but found that using a global object in this case is much easier and doesn't seem dangerous. 
But I understand where I was wrong.

File: AutoRaise.mm [hunk 12]
```diff
@@ -334,6 +334,130 @@ - (void)onTick:(NSNumber *)timerInterval {
...
+#define _delay parameters[@"delay"]
+#define _warpX parameters[@"warpX"]
+#define _warpY parameters[@"warpY"]
+#define _scale parameters[@"scale"]
+#define _warpMouse parameters[@"warpMouse"]
+
+NSArray   *parametersDictionary = @[@"delay", @"warpX", @"warpY", @"scale"];
```
[the remark this answers is not in the document]
Comment (Dimentium) [reply AutoRaise.mm:346]: Hm. I used that as dictionary (list of keys names).

File: AutoRaise.mm [hunk 13]
```diff
@@ -334,6 +334,130 @@ - (void)onTick:(NSNumber *)timerInterval {
...
+@end
+
+@implementation ConfigClass
+
+- (NSString *) getFilePath: (NSString *) filename {
+    filename = [NSString stringWithFormat: @"%@/%@", NSHomeDirectory(), filename];
+    if (not [[NSFileManager defaultManager] fileExistsAtPath: filename]) { filename = @""; }
+    return filename;
+}
+#define fileExist(filename) ([[self getFilePath: filename] length] > 0)
```
[the remark this answers is not in the document]
Comment (Dimentium) [reply AutoRaise.mm:363]: This is exactly a check for the existence of the file.
To determine where to take the configuration from.

File: AutoRaise.mm (hunk 9 above)
Comment (Dimentium) [reply AutoRaise.mm:612]: But we need to check NSDictionary values every time to know - is there needs to go to next configuration method or not.
Isn't it?

File: AutoRaise.mm (hunk 10 above)
[the remark this answers is not in the document]
Comment (sbmpost) [reply AutoRaise.mm:340]: As always there is no absolute right or wrong. Referring to non constants in defines may lead to unexpected behaviour however: for example: what if parameters suddenly points to a different memory location later in the code?

File: AutoRaise.mm (hunk 13 above)
[the remark this answers is not in the document]
Comment (sbmpost) [reply AutoRaise.mm:363]: I thought fileExistsAtPath did that? Or is that insufficient somehow?

File: AutoRaise.mm (hunk 9 above)
Comment (Dimentium) [reply AutoRaise.mm:612]: Yep! 
But it's return NSString and we need to make BOOL for conditions.
And I found that in Objective-C for any non-null string, except 1 YES and true,  return false for [myString boolValue] :/

Comment (sbmpost) [reply AutoRaise.mm:612]: That is one way of doing it. The other is to blindly overwrite the dictionary keys in priority order, something like:


D = FromCmdLine(
      FromOld(
        FromNew(
          Defaults)));

Each method in this chain would accept a dictionary and return a new one, overwriting a value if the key for it is present

File: AutoRaise.mm (hunk 13 above)
[the remark this answers is not in the document]
Comment (Dimentium) [reply AutoRaise.mm:363]: Yep!
But it's return NSString and we need to make BOOL for conditions.
And I found that in Objective-C for any non-null string, except 1 YES and true, return false for [myString boolValue] :/

File: AutoRaise.mm (hunk 10 above)
[the remark this answers is not in the document]
Comment (Dimentium) [reply AutoRaise.mm:340]: ok. good practice - using #define only for constants. taken.

File: AutoRaise.mm (hunk 9 above)
Comment (Dimentium) [reply AutoRaise.mm:612]: Hmm..
But I wanted to do another logic:
if there is *any* arguments from command-line => we should to ignore all config files.

Comment (sbmpost) [reply AutoRaise.mm:612]: The behaviour would indeed be different. And maybe a bit strange for the end user who suddenly is able to provide warpX via the commandline and warpX through a config file. This can however also be done while chaining methods by saying that if there is already one single key present in the previous dictionary, then return that dictionary instead of the new one. Again now absolute right or wrong here ;-)

Comment (Dimentium) [reply AutoRaise.mm:612]: Ok!
I understand that.

Comment (sbmpost) [reply AutoRaise.mm:611]: It is acceptable :-) I might experiment with it myself at some point however. But let us for now focus on getting your PR merged ;-). I think if you double check the other comments and fix what you think needs fixing, then I will test it thoroughly and we merge.

Comment (sbmpost) [reply AutoRaise.mm:611]: Haha, cool :) I am signing out for today, take it easy and thanks for the effort!

File: AutoRaise.mm [hunk 14]
```diff
@@ -475,48 +598,22 @@ CGEventRef eventTapHandler(CGEventTapProxy proxy, CGEventType type, CGEventRef e
...
-        }
-        if (!delayCount) { delayCount = 2; }
-        if (!cursorScale) { cursorScale = 2; }
-
-        printf("\nBy sbmpost(c) 2021, usage:\nAutoRaise -delay <1=%dms> [-warpX <0.5> -warpY <0.5> -scale <2.5>]"
-               "\nv1.8 started with %d ms delay%s", POLLING_MS, delayCount*POLLING_MS, warpMouse ? ", " : "\n");
+        printf("\nv%s by sbmpost(c) 2021, usage:\nAutoRaise -delay <1=%dms> [-warpX <0.5> -warpY <0.5> -scale <2.0>]", VERSION, POLLING_MS);
+        
+        ConfigClass * config = [[ConfigClass alloc]init];
+        [config readCommandLineConfig: argc];
```
Comment (sbmpost) [inline AutoRaise.mm:607]: probably better named: "readConfig", as it could also read from the config from a file.

File: AutoRaise.mm [hunk 15]
```diff
@@ -334,6 +334,129 @@ - (void)onTick:(NSNumber *)timerInterval {
...
+        [self readHiddenConfig];
+    }
+    return;
+}
+
+- (void) readHiddenConfig {
+    // search for dotfiles
+    NSString *hiddenConfigFile = @"";
+    if ([self fileExist: @".AutoRaise"]) { hiddenConfigFile = @".AutoRaise";
+    } else
```
Comment (sbmpost) [inline AutoRaise.mm:414]: can be combined like so: } else if { ... }

File: AutoRaise.mm [hunk 16]
```diff
@@ -334,6 +334,129 @@ - (void)onTick:(NSNumber *)timerInterval {
...
+- (void) readOriginalConfig;
+- (void) readHiddenConfig;
+- (void) validateParameters;
+@end
+
+@implementation ConfigClass
+
+- (NSString *) getFilePath: (NSString *) filename {
+    filename = [NSString stringWithFormat: @"%@/%@", NSHomeDirectory(), filename];
+    if (not [[NSFileManager defaultManager] fileExistsAtPath: filename]) { filename = @""; }
```
Comment (sbmpost) [inline AutoRaise.mm:355]: I understand better now your use cases for getFilePath and fileExist. What you really want, is to get the file path, and if it not exists, return something that indicates that. This can be done with NULL, so you would say:

    filename = NULL;

Then you can drop fileExists, and in the rest of the code do things like:

    NSString * filename = getFilePath(@"some file");
    if (filename) { ... } else { ... }

**Review by sbmpost — COMMENTED** [review]
Some final comments, but I would say after you looked at those, this can be merged.
Well done!

File: AutoRaise.mm (hunk 16 above)
Comment (Dimentium) [reply AutoRaise.mm:355]: > Then you can drop fileExists, and in the rest of the code do things like:
> 
> ```
> NSString * filename = getFilePath(@"some file");
> if (filename) { ... } else { ... }
> ```

Are we talking about [NSString's -boolValue]?

> Declaration
> @property(readonly) BOOL boolValue;
> Discussion
> This property is YES on encountering one of "Y", "y", "T", "t", or a digit 1-9—the method ignores any trailing characters. This property is NO if the receiver doesn’t begin with a valid decimal text representation of a number.
> 

so for
```
NSString * filename = getFilePath(@"some real existing file");
[@"existing file path" -boolValue] => false 
```
:(

It is possible to replace fileExist with condition like this:
`(not [[self getFilePath: delayFile] isEqual: NULL] )`
but it seems less readable. hm?

File: AutoRaise.mm (hunk 15 above)
Comment (Dimentium) [reply AutoRaise.mm:414]: Ok

File: AutoRaise.mm (hunk 14 above)
Comment (Dimentium) [reply AutoRaise.mm:607]: Sure!
